File: src/java/4_BehavioralPatterns/11_Template/TemplateMethodExercise.java
package com.activemesa.behavioral.templatemethod.exercise;

class Creature
{
  public int attack, health;

  public Creature(int attack, int health)
  {
    this.attack = attack;
    this.health = health;
  }
}

abstract class CardGame
{
  public Creature [] creatures;

  public CardGame(Creature[] creatures)
  {
    this.creatures = creatures;
  }

  // return s-1 if no clear winner (both alive or both dead)
  public int combat(int creature1, int creature2)
  {
    Creature first = creatures[creature1];
    Creature second = creatures[creature2];
    hit(first, second);
    hit(second, first);
    boolean firstAlive = first.health > 0;
    boolean secondAlive = second.health > 0;
    if (firstAlive == secondAlive) return -1;
    return firstAlive ? creature1 : creature2;
  }

  // attacker hits other creature
  protected abstract void hit(Creature attacker, Creature other);
}

class TemporaryCardDamageGame extends CardGame
{
  public TemporaryCardDamageGame(Creature[] creatures)
  {
    super(creatures);
  }

  @Override
  protected void hit(Creature attacker, Creature other)
  {
    int oldHealth = other.health;
    other.health -= attacker.attack;
    if (other.health > 0)
      other.health = oldHealth;
  }
}

class PermanentCardDamageGame extends CardGame
{
  public PermanentCardDamageGame(Creature[] creatures)
  {
    super(creatures);
  }

  @Override
  protected void hit(Creature attacker, Creature other)
  {
    other.health -= attacker.attack;
  }
}

File: src/js/4_BehavioralPatterns/10_Template/template_method_exercise.js
class Creature
{
  constructor(attack, health)
  {
    this.attack = attack;
    this.health = health;
  }
}

class CardGame
{
  constructor(creatures)
  {
    this.creatures = creatures;
  }

  // returns index of winner if there's a winner
  // returns -1 if there's no winner (both alive or both dead)
  combat(creature1index, creature2index)
  {
    let first = this.creatures[creature1index];
    let second = this.creatures[creature2index];
    this.hit(first, second);
    this.hit(second, first);
    let firstAlive = first.health > 0;
    let secondAlive = second.health > 0;
    if (firstAlive === secondAlive) return -1;
    return firstAlive ? creature1index : creature2index;
  }

  hit(attacker, defender)
  {
    throw new Error('Please implement this in inheritors');
  }
}

class TemporaryCardDamageGame extends CardGame
{
  constructor(creatures) {
    super(creatures);
  }

  hit(attacker, defender) {
    let oldHealth = defender.health;
    defender.health -= attacker.attack;
    if (defender.health > 0)
      defender.health = oldHealth;
  }
}

class PermanentCardDamageGame extends CardGame
{
  constructor(creatures) {
    super(creatures);
  }

  hit(attacker, defender) {
    defender.health -= attacker.attack;
  }
}

describe('template method', function()
{
  it('impasse test', function()
  {
    let c1 = new Creature(1, 2);
    let c2 = new Creature(1, 2);
    let game = new TemporaryCardDamageGame([c1, c2]);
    expect(game.combat(0, 1)).toEqual(-1);
    expect(game.combat(0, 1)).toEqual(-1);
  });

  it('temporary murder test', function()
  {
    let c1 = new Creature(1, 1);
    let c2 = new Creature(2, 2);
    let game = new TemporaryCardDamageGame([c1, c2]);
    expect(game.combat(0, 1)).toEqual(1);
  });

  it('double murder test', function()
  {
    let c1 = new Creature(2, 2);
    let c2 = new Creature(2, 2);
    let game = new TemporaryCardDamageGame([c1, c2]);
    expect(game.combat(0, 1)).toEqual(-1);
  });

  it('permanent damage death test', function()
  {
    let c1 = new Creature(1, 2);
    let c2 = new Creature(1, 3);
    let game = new PermanentCardDamageGame([c1, c2]);
    expect(game.combat(0, 1)).toEqual(-1);
    expect(game.combat(0, 1)).toEqual(1);
  });
});